### DISCUSSION TITLE
delay when quitting, spurious characters

### DISCUSSION THREAD

**Discussion by nschloe** [Textualize/textual#170] [General]
Sometimes, when letting a Textual app run for a while, there is a long delay for shutting it down, and it'll produce something like
```
q35;35;25M35;36;24M35;37;24M35;38;24M35;39;24M35;40;24M35;39;24M35;37;24M35;33;25M35;29;25M35;24;25M35;19;26M35;14;26M35;10;27M35;5;27M35;2;28M35;5;24M35;10;23M35;14;21M35;19;20M35;23;19M35;27;19M35;30;18M35;33;18M35;35;18M35;36;18M35;36;17M0;36;17M32;35;17M32;34;17M32;33;17M32;32;16M32;31;16M32;30;15M32;29;15M32;28;14M32;28;13M32;27;13M0;27;13m35;27;12M35;27;11M0;27;11M0;27;11m0;27;11M0;27;11mq0;27;11M0;27;11m0;27;11M0;27;11m
```
on the command line after having exited. I'm having a hard time reproducing this reliably, but the following simple app, running for 30 mins, produces the above output on `q`.
```python
from textual.app import App
from textual.widget import Widget

from rich.panel import Panel
from rich import box


class RenderCounter(Widget):
    def on_mount(self):
        self.set_interval(1.0, self.refresh)

    def render(self) -> Panel:
        return Panel("hi", box=box.SQUARE)


class TestApp(App):
    async def on_mount(self) -> None:
        await self.view.dock(RenderCounter(), size=5)

    async def on_load(self, _):
        await self.bind("q", "quit", "quit")


TestApp.run()
```
The problem gets worse on long-running, more complex apps like [tiptop](https://github.com/nschloe/tiptop).

**Comment by willmcgugan** [comment]
I’ve seen this one myself. Not sure what is causing it. If you want to investigate you could put some log calls in the shutdown procedure, to see what takes so long.

**Reply by nschloe** [reply]
I'm on it. `app.on_key()` and `app.shutdown()` are called in short successionm and `app.shutdown()` completes within 0.1 secs, so the delay is either before or after those.

**Reply by nschloe** [reply]
It seems that `await self.animator.stop()` is pretty slow. Investigating further.

**Reply by willmcgugan** [reply]
Interesting! I wonder if there is some object leak there.

**Reply by nschloe** [reply]
In `animator.stop()`, it's `await self._timer_task` that takes a while. Not sure how to debug further.

**Reply by willmcgugan** [reply]
That suggests that that code in _timer.py isn’t exiting when requested.

**Reply by nschloe** [reply]
`self._timer_task` is
```python
asyncio.get_event_loop().create_task(self._timer.run())
```
Why does this even need to be called on `stop()`?

_Edit:_ Looking at `Timer` now.

**Reply by nschloe** [reply]
The issue appear to be in `timer.run()`. The `break` at
```python
if await wait_for(_wait(), max(0, next_timer - monotonic())):
    break
```
only happens after a long delay. I don't understand the code well enough though.

**Reply by nschloe** [reply]
For the record, this has been fixed in https://github.com/Textualize/textual/pull/192.
